### PR TITLE
Removing descriptor set layouts from HAL IR.

### PR DESCRIPTION
**Pull request by benvanik**
* Renamed `push_constants` to `constants` (as there is no longer a `push_constants` API)
* Dropped `#hal.descriptor_set.layout`
* Removed ordinal from `#hal.descriptor_set.binding` (as ordinals are now implicit)
* Renamed `#hal.descriptor_set.binding` to `#hal.pipeline.binding`
* Removed `set` from `hal.interface.binding.subspan`
* Removed `#hal.interface.binding` and the spooky action at a distance `hal.interface.binding` attr now that ordinals are implicit

Progress on #18154.

### CODE REVIEW

**Review by ScottTodd — APPROVED** [review]
> rename Descriptor to Binding (maybe?)

I don't really have an opinion here. Staying close to Vulkan concepts and terminology is nice.

File: compiler/src/iree/compiler/Dialect/HAL/IR/test/command_buffer_folding.mlir [hunk 1]
```diff
@@ -133,3 +131,6 @@ util.func public @fold_buffer_subspan_into_push_descriptor_set(
       ])
   util.return
 }
+
+// DO NOT SUBMIT dispatch
+// DO NOT SUBMIT dispatch indirect (fold into workgroups)
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Dialect/HAL/IR/test/command_buffer_folding.mlir]: Remove before merge (see pre-commit checks: https://github.com/iree-org/iree/actions/runs/10533791489/job/29190363958?pr=18328#step:4:95)

File: compiler/src/iree/compiler/Codegen/LLVMGPU/test/set_transform_strategy_matmul.mlir [hunk 2]
```diff
@@ -43,19 +43,17 @@
...
 ]>
 func.func @matmul_1() {
   %c0 = arith.constant 0 : index
   %cst = arith.constant 0.000000e+00 : f32
-  %0 = hal.interface.binding.subspan layout(#pipeline_layout) set(0) binding(0) alignment(64) offset(%c0) flags(ReadOnly) : !flow.dispatch.tensor<readonly:tensor<2052x2556xf32>>
-  %1 = hal.interface.binding.subspan layout(#pipeline_layout) set(0) binding(1) alignment(64) offset(%c0) flags(ReadOnly) : !flow.dispatch.tensor<readonly:tensor<2556x2052xf32>>
-  %2 = hal.interface.binding.subspan layout(#pipeline_layout) set(0) binding(2) alignment(64) offset(%c0) : !flow.dispatch.tensor<writeonly:tensor<2052x2052xf32>>
+  %0 = hal.interface.binding.subspan layout(#pipeline_layout) binding(0) alignment(64) offset(%c0) flags(ReadOnly) : !flow.dispatch.tensor<readonly:tensor<2052x2556xf32>>
+  %1 = hal.interface.binding.subspan layout(#pipeline_layout) binding(1) alignment(64) offset(%c0) flags(ReadOnly) : !flow.dispatch.tensor<readonly:tensor<2556x2052xf32>>
+  %2 = hal.interface.binding.subspan layout(#pipeline_layout) binding(2) alignment(64) offset(%c0) : !flow.dispatch.tensor<writeonly:tensor<2052x2052xf32>>
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/test/set_transform_strategy_matmul.mlir:56]: Where are we at with trimming these codegen tests down? Should they need all these `hal` ops?

**Review by ScottTodd — APPROVED** [review]
I sorta lost track of some of the diffs between commits here. Sampled a few .mlir files and read through the `HAL/IR/` files and the changes there LGTM.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/test/set_transform_strategy_matmul.mlir (hunk 2 above)
Comment (benvanik) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/test/set_transform_strategy_matmul.mlir:56]: dunno - no, they shouldn't (very few passes do)

**Pull request merged**